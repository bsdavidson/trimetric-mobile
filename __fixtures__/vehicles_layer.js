export const props = {
  filter: ["!=", "non_existing_attribute", "1"],
  vehiclePoints: {
    type: "FeatureCollection",
    features: [
      {
        type: "Feature",
        properties: {
          type: "vehicle",
          icon: "tram",
          vehicle_id: "101",
          route_id: "200",
          stop_id: "7601",
          bearing: 272
        },
        geometry: {type: "Point", coordinates: [-122.67175, 45.526817]}
      },
      {
        type: "Feature",
        properties: {
          type: "vehicle",
          icon: "tram",
          vehicle_id: "536",
          route_id: "200",
          stop_id: "8372",
          bearing: 298
        },
        geometry: {type: "Point", coordinates: [-122.579895, 45.534035]}
      },
      {
        type: "Feature",
        properties: {
          type: "vehicle",
          icon: "tram",
          vehicle_id: "537",
          route_id: "200",
          stop_id: "8347",
          bearing: 165
        },
        geometry: {type: "Point", coordinates: [-122.563545, 45.53022]}
      },
      {
        type: "Feature",
        properties: {
          type: "vehicle",
          icon: "tram",
          vehicle_id: "538",
          route_id: "90",
          stop_id: "10579",
          bearing: 145
        },
        geometry: {type: "Point", coordinates: [-122.59301, 45.58756]}
      }
    ]
  },
  labelsVisible: true
};

export const vehicles = [
  {
    trip: {trip_id: "7918797", route_id: "200"},
    vehicle: {id: "101", label: "Green Line to Clackamas"},
    position: {
      lat: 45.50887,
      lng: -122.68386,
      bearing: 347,
      odometer: 0,
      speed: 0
    },
    current_stop_sequence: 1,
    stop_id: "13140",
    current_status: 1,
    timestamp: 1516746492,
    congestion_level: 0,
    occupancy_status: 0,
    route_type: 0
  },

  {
    trip: {trip_id: "7919096", route_id: "200"},
    vehicle: {id: "537", label: "Green Line to City Ctr"},
    position: {
      lat: 45.435745,
      lng: -122.56782,
      bearing: 350,
      odometer: 0,
      speed: 0
    },
    current_stop_sequence: 1,
    stop_id: "13132",
    current_status: 1,
    timestamp: 1516746481,
    congestion_level: 0,
    occupancy_status: 0,
    route_type: 0
  },

  {
    trip: {trip_id: "7914971", route_id: "90"},
    vehicle: {id: "538", label: "Red Line to Beaverton"},
    position: {
      lat: 45.53185,
      lng: -122.56438,
      bearing: 292,
      odometer: 0,
      speed: 0
    },
    current_stop_sequence: 7,
    stop_id: "8371",
    current_status: 2,
    timestamp: 1516746477,
    congestion_level: 0,
    occupancy_status: 0,
    route_type: 0
  }
];
