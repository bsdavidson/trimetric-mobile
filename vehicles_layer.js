import React, {Component} from "react";
import {Platform, StyleSheet, Text, View} from "react-native";

import Mapbox from "@mapbox/react-native-mapbox-gl";

import {MIN_LABEL_LAYER_ID} from "./App";

// VehiclePoints example:
//
// {
//   type: "FeatureCollection",
//   features: [
//   {
//     geometry: {
//       type: "Point",
//       coordinates: [-122.570694, 45.463062]
//     },
//     properties: {
//       icon: "tram",
//       route_id: "200",
//       type: "vehicle",
//       vehicle_id: "124"
//     },
//     type: "Feature"
//   },
//  ...
// ]}
//
// args:
// filter = {type:"vehicle_id", value: "1234"}

export function VehiclesLayer(vehiclePoints, filter) {
  if (!vehiclePoints) {
    return null;
  }
  return (
    <View>
      <Mapbox.ShapeSource
        id="vehicle_symbols_source"
        images={{
          assets: ["bus", "tram"]
        }}
        shape={vehiclePoints}>
        <Mapbox.SymbolLayer
          id="vehicle_symbols_layer"
          style={mapStyles.vehicleSymbolsLayer}
          belowLayerID={MIN_LABEL_LAYER_ID}
          filter={
            filter ? ["==", filter.type, filter.value] : ["!=", "test", "123"]
          }
        />
      </Mapbox.ShapeSource>

      <Mapbox.ShapeSource id="vehicle_points_source" shape={vehiclePoints}>
        <Mapbox.CircleLayer
          id="vehicle_points_layer"
          maxZoomLevel={13}
          style={mapStyles.vehiclePointsLayer}
          belowLayerID={MIN_LABEL_LAYER_ID}
          filter={
            filter ? ["==", filter.type, filter.value] : ["!=", "test", "123"]
          }
        />
      </Mapbox.ShapeSource>
    </View>
  );
}

const mapStyles = Mapbox.StyleSheet.create({
  vehicleSymbolsLayer: {
    iconImage: "{icon}",
    iconAllowOverlap: true,
    iconOpacity: 1,
    textField: "{route_id}",
    textOffset: [0, -1],
    textHaloColor: "#FFFFFF",
    textHaloWidth: 3,
    iconSize: Mapbox.StyleSheet.camera(
      {
        0: 0,
        9: 0.04,
        13: 0.1,
        18: 0.3
      },
      Mapbox.InterpolationMode.Linear
    )
  },
  vehiclePointsLayer: {
    circleColor: Mapbox.StyleSheet.source(
      {
        0: "#FF00FF",
        3: "#FF000F"
      },
      "routeType",
      Mapbox.InterpolationMode.Categorical
    ),
    circleOpacity: 0.0,
    circleRadius: Mapbox.StyleSheet.camera(
      {
        0: 0,
        9: 2,
        10: 2.4,
        13: 6
      },
      Mapbox.InterpolationMode.Linear
    )
  }
});
