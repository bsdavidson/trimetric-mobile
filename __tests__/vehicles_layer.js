import "react-native";
import React from "react";
import renderer from "react-test-renderer";

import {mapStateToProps, VehiclesLayer} from "../vehicles_layer";
import {props, vehicles} from "../__fixtures__/vehicles_layer.js";

jest.mock("@mapbox/react-native-mapbox-gl");

it("renders VehiclesLayer", () => {
  const vlNull = renderer.create(<VehiclesLayer />).toJSON();
  expect(vlNull).toMatchSnapshot("null_vehiclesLayer");

  const vl = renderer
    .create(
      <VehiclesLayer
        filter={props.filter}
        vehiclePoints={props.vehiclePoints}
        labelsVisible={props.labelsVisible}
      />
    )
    .toJSON();
  expect(vl).toMatchSnapshot("non_null_vehiclesLayer");
});

it("has mapStateToProps", () => {
  // Filter nothing
  let state = mapStateToProps({
    layerVisibility: {vehicleLabels: true, buses: true, trains: true},
    selectedItems: {
      features: []
    },
    vehicles: []
  });
  expect(state).toEqual({
    filter: ["!=", "non_existing_attribute", "1"],
    labelsVisible: true,
    vehiclePoints: {features: [], type: "FeatureCollection"}
  });

  // Filter to vehicleID
  state = mapStateToProps({
    layerVisibility: {vehicleLabels: true, buses: true, trains: true},
    selectedItems: {
      features: [
        {
          properties: {
            type: "vehicle",
            vehicle_id: "101" // matches 1st one in fixture
          },
          geometry: {
            coordinates: [123, 456]
          }
        },
        {
          properties: {
            type: "stop"
          },
          geometry: {
            coordinates: [768, 998]
          }
        }
      ]
    },
    selectedItemIndex: 0,
    vehicles: vehicles
  });
  expect(state).toEqual({
    filter: ["==", "vehicle_id", "101"],
    labelsVisible: true,
    vehiclePoints: {
      features: [
        {
          geometry: {coordinates: [-122.68386, 45.50887], type: "Point"},
          properties: {
            bearing: 347,
            icon: "tram",
            route_id: "200",
            stop_id: "13140",
            type: "vehicle",
            vehicle_id: "101"
          },
          type: "Feature"
        },
        {
          geometry: {coordinates: [-122.56782, 45.435745], type: "Point"},
          properties: {
            bearing: 350,
            icon: "tram",
            route_id: "200",
            stop_id: "13132",
            type: "vehicle",
            vehicle_id: "537"
          },
          type: "Feature"
        },
        {
          geometry: {coordinates: [-122.56438, 45.53185], type: "Point"},
          properties: {
            bearing: 292,
            icon: "tram",
            route_id: "90",
            stop_id: "8371",
            type: "vehicle",
            vehicle_id: "538"
          },
          type: "Feature"
        }
      ],
      type: "FeatureCollection"
    }
  });

  // Filter only buses
  state = mapStateToProps({
    layerVisibility: {vehicleLabels: true, buses: false, trains: true},
    selectedItems: {
      features: []
    },
    selectedItemIndex: 0,
    vehicles: []
  });
  expect(state).toEqual({
    filter: ["!=", "icon", "bus"],
    labelsVisible: true,
    vehiclePoints: {
      features: [],
      type: "FeatureCollection"
    }
  });

  // Filter only trains
  state = mapStateToProps({
    layerVisibility: {vehicleLabels: true, buses: true, trains: false},
    selectedItems: {
      features: []
    },
    selectedItemIndex: 0,
    vehicles: []
  });
  expect(state).toEqual({
    filter: ["!=", "icon", "tram"],
    labelsVisible: true,
    vehiclePoints: {
      features: [],
      type: "FeatureCollection"
    }
  });
});
