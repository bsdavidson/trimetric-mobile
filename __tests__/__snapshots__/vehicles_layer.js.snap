// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`non_null_vehiclesLayer 1`] = `
<View>
  <ShapeSource
    id="vehicle_symbols_source"
    images={
      Object {
        "assets": Array [
          "bus",
          "tram",
        ],
      }
    }
    shape={
      Object {
        "features": Array [
          Object {
            "geometry": Object {
              "coordinates": Array [
                -122.67175,
                45.526817,
              ],
              "type": "Point",
            },
            "properties": Object {
              "bearing": 272,
              "icon": "tram",
              "route_id": "200",
              "stop_id": "7601",
              "type": "vehicle",
              "vehicle_id": "101",
            },
            "type": "Feature",
          },
          Object {
            "geometry": Object {
              "coordinates": Array [
                -122.579895,
                45.534035,
              ],
              "type": "Point",
            },
            "properties": Object {
              "bearing": 298,
              "icon": "tram",
              "route_id": "200",
              "stop_id": "8372",
              "type": "vehicle",
              "vehicle_id": "536",
            },
            "type": "Feature",
          },
          Object {
            "geometry": Object {
              "coordinates": Array [
                -122.563545,
                45.53022,
              ],
              "type": "Point",
            },
            "properties": Object {
              "bearing": 165,
              "icon": "tram",
              "route_id": "200",
              "stop_id": "8347",
              "type": "vehicle",
              "vehicle_id": "537",
            },
            "type": "Feature",
          },
          Object {
            "geometry": Object {
              "coordinates": Array [
                -122.59301,
                45.58756,
              ],
              "type": "Point",
            },
            "properties": Object {
              "bearing": 145,
              "icon": "tram",
              "route_id": "90",
              "stop_id": "10579",
              "type": "vehicle",
              "vehicle_id": "538",
            },
            "type": "Feature",
          },
        ],
        "type": "FeatureCollection",
      }
    }
  >
    <SymbolLayer
      belowLayerID="place-city-sm"
      filter={
        Array [
          "!=",
          "non_existing_attribute",
          "1",
        ]
      }
      id="vehicle_symbols_layer"
      style={
        Array [
          Object {
            "iconAllowOverlap": true,
            "iconImage": "{icon}",
            "iconOpacity": 1,
            "iconSize": "camera_value",
            "textHaloColor": "#FFFFFF",
            "textHaloWidth": 3,
            "textOffset": Array [
              0,
              -1,
            ],
            "textOptional": true,
          },
          Object {
            "textField": "{route_id}",
          },
        ]
      }
    />
  </ShapeSource>
</View>
`;

exports[`null_vehiclesLayer 1`] = `null`;
